### DISCUSSION TITLE
Windows Terminal replacement

### DISCUSSION THREAD

**Discussion by DRSDavidSoft** [cmderdev/cmder#2864] [Polls]
### Description

Hi there! 👋🏻

As you know, **ConEmu** is the terminal emulator that has always been used in Cmder to provide the underlying shells (such as Clink, Powershell, Bash) from the start.  This has been great as ConEmu was a feature-rich and awesome replacement to the old, Console Host (conhost.exe) in Windows XP, Vista, 7.

Cmder is also compatible with other terminal emulators, such as **[Windows Terminal](https://github.com/cmderdev/cmder/wiki/Seamless-Windows-Terminal-Integration)**, which is finally not only a good, but a **great** terminal emulator from Microsoft.

There are a variety of other terminal emulators that you can use with Cmder, which are explained in our [Wiki](https://github.com/cmderdev/cmder/wiki).

Do you personally prefer ConEmu, Windows Terminal, or any other terminal emulators?

Would you support that Cmder would **officially** migrate from ConEmu to Windows Terminal?

We'd appreciate hearing your thoughts on this. **:smile: Please submit your vote at the bottom of this message!**

**👉 Update:** Windows Terminal support is being built into Cmder project, see #2877 for more

https://github.com/cmderdev/cmder/assets/4673812/79051ff3-3a7b-4922-8649-5dbfb2d63ec3

**Comment by funlockz** [comment]
Although it is already possible with the init script e.g. putting something like this in WT
`cmd.exe /k %CMDER_ROOT%/vendor/init.bat`
or if you want to use with PS in your WT:
`"C:\Program Files\PowerShell\7\pwsh.exe" -ExecutionPolicy Bypass -NoLogo -NoProfile -NoExit -Command "Invoke-Expression 'Import-Module ''%CMDER_ROOT%\vendor\profile.ps1'''"`

For me it seems to work just just fine. The only thing I noticed is that if you want to register custom scripts with Clink, it is shared across all sessions, so multiple different profiles in WT will share the same scripts (e.g. if you enable starship or ohmyposh in one it will appear in all) but that is not a big issue so far.

**Reply by DRSDavidSoft** [reply]
Of course, you can add the dedicated Cmder profile to Windows Terminal:
```
        {
            "guid": "{00000000-0000-0000-ba54-000000000132}",
            "commandline": "cmd.exe /k \"%cmder_root%/vendor/init.bat\"",
            "icon": "%cmder_root%/icons/cmder.ico",
            "name": "Cmder", // You can give a custom name here
            "closeOnExit": "graceful",
            "hidden": false
        },
```

Taken from the [wiki page](https://github.com/cmderdev/cmder/wiki/Seamless-Windows-Terminal-Integration).

**Reply by candrapersada** [reply]
will there be a feature to will automatically find Cmder and show it in the shell settings like Tabby in Windows Terminal?

**Reply by DRSDavidSoft** [reply]
@candrapersada If we replace ConEmu with Windows Terminal, we will work something out like this. Currently, @daxgames has written some PowerShell scripts that would add the Cmder profile permanently to Windows Terminal.

Tabby's auto-detection is a neat feature. I hope something like that would also be added to WT.

**Reply by candrapersada** [reply]
> Tabby's auto-detection is a neat feature. I hope something like that would also be added to WT.

yes

**Comment by daxgames** [comment]
First looks at Windows Terminal Integration

![image](https://github.com/cmderdev/cmder/assets/7318053/4567d669-4613-4fc2-92d2-7dc415255a05)

So far I have:

- Added an argument to build.ps1 that allows building with Windows terminal instead of Conemu.
- Created a '%cmder_root%\vendor\WindowsTerminal_settings_default.json' that results in the WindowsTerminal Session and profiles shown above.
  - All sessions work as they do today in Cmder using Conemu.
  - Also includes changes to the word separator settings so it works more m=like we would expect it ti work.
- Modified Cmder.bat to launch windows terminal by default if it exists and put the default config file where it needs to be before launching Windows Terminal

**Reply by candrapersada** [reply]
Is there an option to import or export settings from Windows terminal cmder to normal Windows terminal?

**Reply by daxgames** [reply]
This is 'normal' Windows Terminal running in portable mode.  The only difference is the Cmder profiles that are configured/running in it.

**Reply by candrapersada** [reply]
what I mean is how to use Cmder profiles or cmder settings in wt system
`%LOCALAPPDATA%\Microsoft\WindowsApps\wt`

**Reply by daxgames** [reply]
You would need to create the Cmder profiles in Windows Terminal if you're not using the embedded Windows Terminal that comes with Cmder.

As far as I know, windows terminal doesn't offer an import export feature.

**Reply by candrapersada** [reply]
Backup Windows Terminal
```
copy /y "%LOCALAPPDATA%\Packages\Microsoft.WindowsTerminal_8wekyb3d8bbwe\LocalState\settings.json" "%LOCALAPPDATA%\Packages\Microsoft.WindowsTerminal_8wekyb3d8bbwe\LocalState\settings_backup.json"
```
import or export settings from Windows Terminal Cmder to Windows Terminal
```
copy /y "%cmder_root%\vendor\windows_terminal_default_settings.json" "%LOCALAPPDATA%\Packages\Microsoft.WindowsTerminal_8wekyb3d8bbwe\LocalState\settings.json"
```
Restore Windows Terminal
```
copy /y "%LOCALAPPDATA%\Packages\Microsoft.WindowsTerminal_8wekyb3d8bbwe\LocalState\settings_backup.json" "%LOCALAPPDATA%\Packages\Microsoft.WindowsTerminal_8wekyb3d8bbwe\LocalState\settings.json"
```

**Reply by bossanovaorca** [reply]
could you share the dedicated Cmder profiles? so far I've only seen the default cmd and powershell profiles from the [wiki](https://github.com/cmderdev/cmder/wiki/Seamless-Windows-Terminal-Integration)

**Reply by daxgames** [reply]
@bossanovaorca https://github.com/cmderdev/cmder/blob/43a5e7e0b15584b9f128a56b39b3b1f21e1dba4f/vendor/windows_terminal_default_settings.json

**Comment by RaXorX** [comment]
How about CMDer with Tabby?

**Reply by DRSDavidSoft** [reply]
**✔️ You can use Cmder with Tabby with this official guide:**
https://github.com/cmderdev/cmder/wiki/Seamless-Terminus-Integration

TL:DR; just make sure the `CMDER_ROOT` environment is set to the installation root of Cmder. Tabby will auto-detect the Cmder shell.

**Comment by candrapersada** [comment]
is it possible to use `%WT_SETTINGS_DIR%` for `windows terminal` installed from the Microsoft Store without using `windows terminal portablel`+`Cmder.exe`?
![image](https://github.com/user-attachments/assets/805393e0-3e18-4127-8b03-44721c99d002)

**Reply by daxgames** [reply]
See: https://github.com/microsoft/terminal/discussions/17191

I dont believe it is just for portable mode.

**Reply by candrapersada** [reply]
is there any support or not `windows terminal` nonportable will auto-detect the `Cmder` shell?

**Reply by DRSDavidSoft** [reply]
@candrapersada The Windows Terminal team have been working on an alternative approach for this request. Please see:

https://github.com/microsoft/terminal/issues/16371

**Reply by candrapersada** [reply]
but until now still not told how to Automatic profile detection for Cmder?
